### ISSUE TITLE
Fix Daily Archive.org Import Query & Backfill ~100k imports from `printdisabled`

### ISSUE THREAD

**Issue by mekarpeles** [internetarchive/openlibrary#732]
See: https://github.com/internetarchive/openlibrary/wiki/archive.org-%E2%86%94-Open-Library-synchronisation

## Subtasks
- [x] #1047 Find out why `printdisabled` items not imported into Open Library
- [ ] #1046 Switch Archive.org items to use `openlibrary_edition` instead of `openlibrary` keys

## Scope

As of 2025-02-18:

There are 101,880 records
`collection:internetarchivebooks AND collection:printdisabled AND -collection:inlibrary AND -openlibrary_edition:* AND isbn:* AND format:MARC`

https://archive.org/search?query=collection%3Ainternetarchivebooks+AND+collection%3Aprintdisabled+AND+-collection%3Ainlibrary+AND+-openlibrary_edition%3A*+AND+isbn%3A*+AND+format%3AMARC

`collection:internetarchivebooks AND collection:printdisabled AND -collection:inlibrary AND -openlibrary_edition:*`

, incognito there are 85,548 matches for the query `collection:printdisabled AND -collection:inlibrary AND -openlibrary_edition:*`: https://archive.org/search?query=collection%3Aprintdisabled+AND+-collection%3Ainlibrary+AND+-openlibrary_edition%3A*

Of these:
- 62,783 have `isbn:*`
- 38,978 have `format:MARC`
- 35,351 have `format:MARC AND isbn:*`

## Background

consider which of these unlinked Archive.org items should be used to create or update Open Library editions.

As of #549, we have a privileged OL admin endpoint which can write back `openlibrary_work` and `openlibrary_edition` to Archive.org items. We also have an endpoint (i.e. the one used for "add to Staff Picks" btn on works pages) to add `openlibrary_subject`.

**Comment by mekarpeles** [comment]
Related is #731

**Comment by hornc** [comment]
I am currently syncing 107488 OL `In Library` items with OCAIDs that do not have `openlibrary_edition` populated. 

The list was generated using an ia-client query, and cross-referencing the ocaids against the OL edition dumps of Jan 2018.


I have identified ~14K `In Library` items that need to be synced but don't have works. Half of those have no authors linked, and of the ones that do, some are incorrect due to un-reverted bad author merges. It appears not having a work broke the merge revert process. Further investigation needed as part of resolving the orphaned editions problem.

**Comment by hornc** [comment]
Querying IA for [`In Library` items without `openlibrary_edition`](https://archive.org/advancedsearch.php?q=collection%3Ainlibrary+AND+%21openlibrary_edition%3A*&fl%5B%5D=identifier&sort%5B%5D=&sort%5B%5D=&sort%5B%5D=&rows=50&page=1&output=json) now gives 57625 results. I believe these are now either items not on OL, or orphaned editions that need works before syncing (see #792)

**Comment by mekarpeles** [comment]
As a one-off, I think this is done. Moving forward (hopefully) ocaids should be getting written back by importbot.

**Comment by mekarpeles** [comment]
@hornc, what do you say we take a stab at this?

Out of the 697,460 books on archive.org `inlibrary` 88,712 are not sync'ed with `openlibrary_edition` and `openlibrary_work`.

As an administrator, if we do the same query against `printdisabled` (and perhaps add an additional argument of `isbn:*` to reduce the noice), we see 600,932 items on archive.org which potentially aren't on OL or aren't linked:

https://archive.org/search.php?query=collection%3Aprintdisabled%20AND%20%21openlibrary_edition%3A%2A%20AND%20isbn%3A%2A

In short, there's an opportunity to either create or/and link at least 88k `inlibrary` items, and potentially ~600-800k `printdisabled`books.

**Comment by hornc** [comment]
@mekarpeles Curiouser and curiouser, the first example from the `inlibrary` version of the query https://archive.org/search.php?query=collection%3Ainlibrary%20AND%20%21openlibrary_edition%3A%2A%20AND%20isbn%3A%2A
is already on OL
https://archive.org/details/isbn_9780312604530
https://openlibrary.org/books/OL24643499M/The_Borgia_Betrayal
and attempting to sync it fails repeatedly: 
```
{
error: "borgiabetrayal00sara failed: {"success":false,"error":"no changes to _meta.xml"}"
}
```
This may be an example of what I was asking for to confirm #950 
I'll keep :eyes:

**Comment by hornc** [comment]
gah! there are two scans of this book on archive.org, https://archive.org/details/isbn_9780312604530 which is not linked to the OL item, and https://archive.org/details/borgiabetrayal00sara which is.

The error message above is accurate when it says `no changes`, but it's not actually a "failure" as such.   I don't know how common these duplicates are, still investigating.

**Comment by hornc** [comment]
A subset of the 88K have the old style `openlibrary` field, 32K:  https://archive.org/search.php?query=collection%3Ainlibrary%20AND%20%21openlibrary_edition%3A%2A%20AND%20isbn%3A%2A%20AND%20openlibrary%3A%2A So we know they are on open library.

Some of those appear to be orphaned editions, so the writeback fails with ```{
error: "No qualifying edition"
}```  which is slightly misleading (it's the work that is missing, not edition),  others do not have a ocaid on the edition. I plan to script something that checks whether the edition is an orphan, adds the ocaid if needed, and triggers the writeback. Then we can investigate the remaining items (I was planning to do ISBN lookups, but fixing the ones that have know OLIDs is step 1)

**Comment by mekarpeles** [comment]
@hornc is this wrapper up? Can the checkbox be clicked and the issue closed?

**Comment by hornc** [comment]
Breakdown and details at https://github.com/internetarchive/openlibrary/wiki/archive.org-%E2%86%94-Open-Library-synchronisation

**Comment by mekarpeles** [comment]
Almost 700,000 items need to be synced. Success & re-closing looks like an import run (https://openlibrary.org/admin/imports) over these IDs.
https://archive.org/search.php?query=collection%3Ainlibrary%20AND%20%21openlibrary_work%3A%2A

**Comment by mekarpeles** [comment]
**Problem(s):**
* IA metadata can sometimes be wrong (e.g. valid but mismatched MARC record)
* OpenLibrary availability buttons are significantly incorrect in our current model
* There may be significant drift / mis-sync between IA<->OL records

**Proposal:**
Can we once a month:
* export a set of OpenLibrary.org's ocaid, work_key, edition_key , isbn, title from the dump
* export Archive.org's inlibrary & (printdisabled AND !collection:inlibrary) independently w/ ocaid, openlibrary_work, openlibrary_edition, isbn, language, title

And:
* Identify & flag matching IA/OL records w/ conflicting metadata (title, isbn, IA <-> OL keys)
* Update IA item back-links to OL
* Update availability type for OL solr

**Comment by mekarpeles** [comment]
**Resources**
* https://docs.google.com/document/d/1d0os8BLLt4j_mn36W05oZYFbQToNhD2Gg2bUEmpm8eQ/edit#heading=h.eqrdpmbrlr48

@cdrini @scottbarnes

**Comment by mekarpeles** [comment]
See also:
* #7356
* #7217

**Comment by mekarpeles** [comment]
@scottbarnes and @cdrini to evaluate whether new script/query addresses drift and, if so, replace the get_candidates query in core/ia.py with the better query.

**Comment by mekarpeles** [comment]
8.5M items
```
mediatype:texts AND format:pdf AND collection:(internetarchivebooks OR toronto OR europeanlibraries OR americana) AND -issn:* -collection:(opensource OR additional_collections OR periodicals OR serials OR bplgovdocs OR microfilm OR microfiche OR dlarc OR spiritualfrontiersarchive OR industrydocuments OR *amphlet* OR peel_newspapers OR fedlink OR newmannumismatic) -subject:(*earbook* OR *vertissement* OR *agazine* OR *ewsletter OR Logbooks OR *posters* OR *ewspaper* OR periodicals OR  *District Court OR *Documents* OR Personnel* OR microfilm OR United States Politics and Government*) -creator:(united states* OR "u.s. mint" OR *Court*) -volume:* AND scanningcenter:* 
```

2.6M items have no `format:MARC`

**Comment by mekarpeles** [comment]
Closed by #10704
